codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/BuilderGenerator.kt

```diff
@@ -76,10 +80,59 @@ abstract class BuilderCustomization : NamedCustomization<BuilderSection>()
 fun RuntimeConfig.operationBuildError() = RuntimeType.operationModule(this).resolve("error::BuildError")
 fun RuntimeConfig.serializationError() = RuntimeType.operationModule(this).resolve("error::SerializationError")
 
+fun MemberShape.enforceRequired(
+    field: Writable,
+    codegenContext: CodegenContext,
+    produceOption: Boolean = true,
+): Writable {
+    if (!this.isRequired) {
+        return field
+    }
+    val shape = this
+    val ctx = arrayOf(
+        "checkSetString" to checkSetString,
+        "error" to OperationBuildError(codegenContext.runtimeConfig).missingField(
+            codegenContext.symbolProvider.toMemberName(shape), "A required field was not set",
+        ),
+        "field" to field,
+    )
+    val unwrapped = when (codegenContext.model.expectShape(this.target)) {
+        is StringShape -> writable {
+            rustTemplate(
+                "#{checkSetString}(#{field}).ok_or_else(||#{error})?",
+                *ctx,
+            )
+        }
+
+        else -> writable {
+            rustTemplate("#{field}.ok_or_else(||#{error})?", *ctx)
+        }
+    }
+    return unwrapped.letIf(produceOption) { writable { rust("Some(#T)", it) } }
+}
+
+private val checkSetString = RuntimeType.forInlineFun("non_empty_str", RustModule.private("serde_util")) {
+    rustTemplate(
+        """
+        pub (crate) fn non_empty_str<T: AsRef<str>>(field: Option<T>) -> Option<T> {
+            if let Some(field) = field {
+                if field.as_ref() != "" {
+                    return Some(field)
+                }
+            }
+            None
```

Comment:
I think this can be one line:
```suggestion
field.map(AsRef::as_ref).filter(|s| s != "")
```